### PR TITLE
Added support for GitHub pages using docfx

### PR DESCRIPTION
**Pull request by jscarle**
This PR adds support for publishing documentation as [GitHub pages](https://pages.github.com/) using [docfx](https://dotnet.github.io/docfx/).

**After** this PR is merged, three things need to be done for setup to be completed.

1) GitHub pages must be activated for this repository by navigating to Settings => Pages and changing the Build and deployment source to GitHub Actions here (https://github.com/sshnet/SSH.NET/settings/pages)
image

2) From the main repository page (https://github.com/sshnet/SSH.NET) on the About gear for the settings:
image

The option Use your GitHub Pages website should be checked:
image

3) The GitHub Action for Docs must be manually run at least once (https://github.com/sshnet/SSH.NET/actions/workflows/docs.yml):
image

Once the docs have been published once, they will be _automatically_ published everytime develop is merged with **master**.

### CODE REVIEW

**Comment by jscarle** [comment]
@WojciechNagorski Ready for review and merge. Let me know if you have any questions!

**Comment by jscarle** [comment]
@WojciechNagorski @Rob-Hague This is now ready to merge and synced with develop.

File: src/Renci.SshNet/Common/AsyncResult`1.cs [hunk 1]
```diff
@@ -6,7 +6,9 @@ namespace Renci.SshNet.Common
     /// Base class to encapsulates the results of an asynchronous operation that returns result.
     /// </summary>
     /// <typeparam name="TResult">The type of the result.</typeparam>
+#pragma warning disable SA1649 // Normal convention is to name generic files `1, naming this AsyncResult{TResult}.cs breaks docfx.
     public abstract class AsyncResult<TResult> : AsyncResult
+#pragma warning restore SA1649
```
Comment (Rob-Hague) [inline src/Renci.SshNet/Common/AsyncResult`1.cs]: It looks like you renamed the file, so presumably the suppression is not needed now

```suggestion
    public abstract class AsyncResult<TResult> : AsyncResult
```

**Review by Rob-Hague — COMMENTED** [review]
Thanks @jscarle. I don't know anything about it but it looks nice and simple. Two questions:

1. Is it possible to preview the output without merging, either locally or on GH?
2. Is it possible to have the additional docs (features.md, foreward.md, usage.md) without duplicating from the README? E.g. including snippets of markdown (or something). If not, can we just directly use the README for now to avoid having to update in multiple places?

File: src/Renci.SshNet/Common/AsyncResult`1.cs (hunk 1 above)
Comment (jscarle) [reply src/Renci.SshNet/Common/AsyncResult`1.cs]: If you try it, you'll notice that as soon as you remove the warning, it'll tell you to rename the file back to `AsyncResult{TResult}.cs` (which I've never seen to be honest), and it'll break docfx.

**Comment by jscarle** [comment]
> Thanks @jscarle. I don't know anything about it but it looks nice and simple. Two questions:
> 
> 1. Is it possible to preview the output without merging, either locally or on GH?

Certainly! From the root of the repository, simply run the two commands from the pipeline:

```cmd
dotnet tool update -g docfx
docfx ./docfx/docfx.json
```

That will produce the site in `./docfx/_site`, which as you'll see, is just a plain HTML site. You can view it by opening `index.html`.

> 2. Is it possible to have the additional docs (features.md, foreward.md, usage.md) without duplicating from the README? E.g. including snippets of markdown (or something). If not, can we just directly use the README for now to avoid having to update in multiple places?

Sure, we can do whatever we'd like with the additional content of the site. I just split up the README as a starting point. I could also work into a way to include the README directly. Though the objective would be to simplify the README and instead grow this site.

File: src/Renci.SshNet/Common/AsyncResult`1.cs (hunk 1 above)
Comment (Rob-Hague) [reply src/Renci.SshNet/Common/AsyncResult`1.cs]: Right. I assumed it had been telling you to rename to `1. This fixes it:

```diff
diff --git a/stylecop.json b/stylecop.json
index ff7c9dbf..ff643f45 100644
--- a/stylecop.json
+++ b/stylecop.json
@@ -3,7 +3,9 @@
     "settings": {
         "documentationRules": {
             "xmlHeader": false,
-            "documentInternalElements": false
+            "documentInternalElements": false,
+            "fileNamingConvention": "metadata"
         },
         "layoutRules": {
             "newlineAtEndOfFile": "require"
```

Comment (jscarle) [reply src/Renci.SshNet/Common/AsyncResult`1.cs]: Indeed it did, thanks.

Comment (jscarle) [reply src/Renci.SshNet/Common/AsyncResult`1.cs]: I removed the warning as well.

**Comment by Rob-Hague** [comment]
On index.html I don't see anything to click on to get to the code. Does that get hooked up automatically?

For the additional content, I'll leave it up to you if you expect the site and the readme to diverge in the near future. If they are going to stay equal for a while, then having just one place to update would be nicer from a maintenance perspective if it's easy to do.

**Comment by jscarle** [comment]
> On index.html I don't see anything to click on to get to the code. Does that get hooked up automatically?

Sorry about that, I forgot. You have to do:
```batch
docfx ./docfx/docfx.json --serve
```
Then, at the end you should see:
```batch
Serving "E:\github\SSH.NET\docfx\_site" on http://localhost:8080. Press Ctrl+C to shut down.
```
It runs a little bit of JavaScript, so that's probably why it won't completely load locally without running serve.

**Review by Rob-Hague — APPROVED** [review]

**Comment by Rob-Hague** [comment]
http://sshnet.github.io/SSH.NET/

**Comment by jscarle** [comment]
Good stuff!

**Comment by scott-xu** [comment]
Need to update https://github.com/sshnet/SSH.NET/blob/develop/docfx/docs/foreword.md/

**Review by Buildervt — COMMENTED** [review]
Emmy88724080.$

**Pull request merged**